File: app/src/features/tasks/services/task.service.ts
/**
 * Task Service
 * 
 * Business logic layer for task operations
 */

import { TaskRepository, TaskHierarchy } from '@/core/database/repositories'
import {
    Task,
    CreateTaskDto,
    UpdateTaskDto,
    TaskStats,
    CreateTaskDtoSchema,
    UpdateTaskDtoSchema
} from '../types/task.types'

const taskRepository = new TaskRepository()

export class TaskService {
    /**
     * Create a new task
     */
    static async create(data: CreateTaskDto): Promise<Task> {
        // Validate input
        const validated = CreateTaskDtoSchema.parse(data)
        return await taskRepository.create(validated)
    }

    /**
     * Get all tasks for a project
     */
    static async getAll(projectId: string): Promise<Task[]> {
        return await taskRepository.findByProjectId(projectId)
    }

    /**
     * Get root tasks (no parent) for a project
     */
    static async getRootTasks(projectId: string): Promise<Task[]> {
        return await taskRepository.findRootTasks(projectId)
    }

    /**
     * Get subtasks for a parent task
     */
    static async getSubtasks(parentId: string): Promise<Task[]> {
        return await taskRepository.findSubtasks(parentId)
    }

    /**
     * Get task by ID
     */
    static async getById(id: string): Promise<Task | null> {
        return await taskRepository.findById(id)
    }

    /**
     * Get task hierarchy
     */
    static async getHierarchy(id: string): Promise<TaskHierarchy> {
        return await taskRepository.getHierarchy(id)
    }

    /**
     * Update task
     */
    static async update(id: string, data: UpdateTaskDto): Promise<Task> {
        // Validate input
        const validated = UpdateTaskDtoSchema.parse(data)
        return await taskRepository.update(id, validated)
    }

    /**
     * Delete task and all subtasks
     */
    static async delete(id: string): Promise<void> {
        return await taskRepository.delete(id)
    }

    /**
     * Move task to a different parent
     */
    static async move(id: string, newParentId: string | null): Promise<Task> {
        return await taskRepository.moveTask(id, newParentId)
    }

    /**
     * Reorder task
     */
    static async reorder(id: string, newOrder: number): Promise<Task> {
        return await taskRepository.reorder(id, newOrder)
    }

    /**
     * Get tasks by status
     */
    static async getByStatus(
        projectId: string,
        status: 'todo' | 'in_progress' | 'done'
    ): Promise<Task[]> {
        return await taskRepository.findByStatus(projectId, status)
    }

    /**
     * Get tasks by priority
     */
    static async getByPriority(
        projectId: string,
        priority: 'low' | 'medium' | 'high'
    ): Promise<Task[]> {
        return await taskRepository.findByPriority(projectId, priority)
    }

    /**
     * Search tasks
     */
    static async search(projectId: string, query: string): Promise<Task[]> {
        return await taskRepository.search(projectId, query)
    }

    /**
     * Get task statistics for a project
     */
    static async getStats(projectId: string): Promise<TaskStats> {
        const allTasks = await taskRepository.findByProjectId(projectId)
        const now = new Date()

        return {
            total: allTasks.length,
            todo: allTasks.filter(t => t.status === 'todo').length,
            inProgress: allTasks.filter(t => t.status === 'in_progress').length,
            done: allTasks.filter(t => t.status === 'done').length,
            overdue: allTasks.filter(t =>
                t.status !== 'done' &&
                t.dueDate &&
                new Date(t.dueDate) < now
            ).length,
        }
    }

    /**
     * Complete a task
     */
    static async complete(id: string): Promise<Task> {
        return await taskRepository.update(id, {
            status: 'done',
            completedAt: new Date()
        })
    }

    /**
     * Reopen a completed task
     */
    static async reopen(id: string): Promise<Task> {
        return await taskRepository.update(id, {
            status: 'todo',
            completedAt: null
        })
    }

    /**
     * Create a subtask
     */
    static async createSubtask(parentId: string, data: Omit<CreateTaskDto, 'parentId'>): Promise<Task> {
        const parentTask = await taskRepository.findById(parentId)
        if (!parentTask) {
            throw new Error('Parent task not found')
        }

        return await this.create({
            ...data,
            projectId: parentTask.projectId,
            parentId,
        })
    }

    /**
     * Get overdue tasks
     */
    static async getOverdue(projectId: string): Promise<Task[]> {
        const allTasks = await taskRepository.findByProjectId(projectId)
        const now = new Date()

        return allTasks.filter(t =>
            t.status !== 'done' &&
            t.dueDate &&
            new Date(t.dueDate) < now
        )
    }

    /**
     * Bulk update task order
     */
    static async bulkReorder(updates: { id: string; order: number }[]): Promise<void> {
        for (const update of updates) {
            await taskRepository.reorder(update.id, update.order)
        }
    }
}

File: app/src-tauri/src/services/task_service.rs
use crate::error::{AppError, AppResult};
use crate::models::{CreateTaskDto, Task, UpdateTaskDto, TaskWithChildren};
use uuid::Uuid;

/// Task service for business logic
pub struct TaskService;

impl TaskService {
    /// Create a new task
    pub async fn create_task(data: CreateTaskDto) -> AppResult<Task> {
        // Validate input
        if data.title.is_empty() {
            return Err(AppError::InvalidInput("Task title cannot be empty".into()));
        }

        if data.project_id.is_empty() {
            return Err(AppError::InvalidInput("Project ID cannot be empty".into()));
        }

        let now = chrono::Utc::now().timestamp();
        let task = Task {
            id: Uuid::new_v4().to_string(),
            project_id: data.project_id,
            parent_id: data.parent_id,
            title: data.title,
            description: data.description,
            status: data.status.unwrap_or_else(|| "todo".to_string()),
            priority: data.priority.unwrap_or_else(|| "medium".to_string()),
            due_date: data.due_date,
            completed_at: None,
            created_at: now,
            updated_at: now,
            order: data.order.unwrap_or(0),
            tags: data.tags,
        };

        // TODO: Save to database via IPC to frontend repository

        Ok(task)
    }

    /// Get all tasks for a project
    pub async fn list_tasks(project_id: String) -> AppResult<Vec<Task>> {
        if project_id.is_empty() {
            return Err(AppError::InvalidInput("Project ID cannot be empty".into()));
        }

        // TODO: Fetch from database
        Ok(vec![])
    }

    /// Get root tasks (no parent) for a project
    pub async fn list_root_tasks(project_id: String) -> AppResult<Vec<Task>> {
        if project_id.is_empty() {
            return Err(AppError::InvalidInput("Project ID cannot be empty".into()));
        }

        // TODO: Fetch from database
        Ok(vec![])
    }

    /// Get subtasks for a parent task
    pub async fn list_subtasks(parent_id: String) -> AppResult<Vec<Task>> {
        if parent_id.is_empty() {
            return Err(AppError::InvalidInput("Parent ID cannot be empty".into()));
        }

        // TODO: Fetch from database
        Ok(vec![])
    }

    /// Get task by ID
    pub async fn get_task(id: String) -> AppResult<Task> {
        if id.is_empty() {
            return Err(AppError::InvalidInput("Task ID cannot be empty".into()));
        }

        // TODO: Fetch from database
        Err(AppError::NotFound("Task", id))
    }

    /// Get task hierarchy (task with all descendants)
    pub async fn get_task_hierarchy(id: String) -> AppResult<TaskWithChildren> {
        if id.is_empty() {
            return Err(AppError::InvalidInput("Task ID cannot be empty".into()));
        }

        // TODO: Fetch from database with hierarchy
        Err(AppError::NotFound("Task", id))
    }

    /// Update task
    pub async fn update_task(id: String, data: UpdateTaskDto) -> AppResult<Task> {
        if id.is_empty() {
            return Err(AppError::InvalidInput("Task ID cannot be empty".into()));
        }

        // TODO: Update in database
        Err(AppError::NotFound("Task", id))
    }

    /// Delete task (and all subtasks)
    pub async fn delete_task(id: String) -> AppResult<()> {
        if id.is_empty() {
            return Err(AppError::InvalidInput("Task ID cannot be empty".into()));
        }

        // TODO: Delete from database (cascade to subtasks)
        Err(AppError::NotFound("Task", id))
    }

    /// Move task to a different parent
    pub async fn move_task(id: String, new_parent_id: Option<String>) -> AppResult<Task> {
        if id.is_empty() {
            return Err(AppError::InvalidInput("Task ID cannot be empty".into()));
        }

        // TODO: Update parent_id in database
        Err(AppError::NotFound("Task", id))
    }

    /// Reorder task
    pub async fn reorder_task(id: String, new_order: i32) -> AppResult<Task> {
        if id.is_empty() {
            return Err(AppError::InvalidInput("Task ID cannot be empty".into()));
        }

        // TODO: Update order in database
        Err(AppError::NotFound("Task", id))
    }

    /// Get tasks by status
    pub async fn list_tasks_by_status(project_id: String, status: String) -> AppResult<Vec<Task>> {
        if project_id.is_empty() {
            return Err(AppError::InvalidInput("Project ID cannot be empty".into()));
        }

        // Validate status
        if !["todo", "in_progress", "done"].contains(&status.as_str()) {
            return Err(AppError::InvalidInput(format!(
                "Invalid status '{}'. Must be one of: todo, in_progress, done",
                status
            )));
        }

        // TODO: Fetch from database
        Ok(vec![])
    }

    /// Search tasks
    pub async fn search_tasks(project_id: String, query: String) -> AppResult<Vec<Task>> {
        if project_id.is_empty() {
            return Err(AppError::InvalidInput("Project ID cannot be empty".into()));
        }

        // TODO: Search in database
        Ok(vec![])
    }
}
